pandas/core/algorithms.py

```diff
@@ -430,6 +430,12 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> np.ndarray:
         # handle categoricals
         return cast("Categorical", comps).isin(values)
 
+    if needs_i8_conversion(comps):
+        # Dispatch to DatetimeLikeIndexMixin.isin
+        from pandas import Index
+
+        return Index(comps).isin(values)
```

Comment:
are you sure?

ensure_data and below does this

Reply:
the way we call ensure_data ends up casting dt64 to int64

Reply:
why can't you fix it in _ensure_data then, I am not averse to dispatching to the EA itself, which is what we are already doing for Categorical, or is that the plan. If that is the plan, then I would think you could actually remove all off this and simply dispatch (e.g. most of this logic then would actually be handled by PandasArray or the EA as appropriate).

Reply:
There are a bunch of GH issues about isin (added a new label yesterday), many of which are about unwanted casting.  I haven't looked at them all closely tet.

> why can't you fix it in _ensure_data then

Even if we fix that (which yes, i'll take a look, but probably in a separate PR), we still need to dispatch to the `(Period|Timedelta|Datetime)Index` to get correct casting of the `other` (assuming we want `algos.isin(dtlike, other)` to match `dtlike.isin(other)`, which I for one do)

>  am not averse to dispatching to the EA itself, which is what we are already doing for Categorical, or is that the plan

ATM Categorical has `isin` but thats it.  #20617 suggests adding isin to EA, but thats definitely outside the scope of this PR.